apps/aeoracle/src/aeo_register_tx.erl

```diff
@@ -0,0 +1,179 @@
+query_spec(#oracle_register_tx{query_spec = QuerySpec}) ->
+    QuerySpec.
+
+-spec response_spec(register_tx()) -> aeo_oracles:type_spec().
+response_spec(#oracle_register_tx{response_spec = ResponseSpec}) ->
+    ResponseSpec.
+
+-spec query_fee(register_tx()) -> integer().
+query_fee(#oracle_register_tx{query_fee = QueryFee}) ->
+    QueryFee.
+
+-spec ttl(register_tx()) -> aeo_oracles:ttl().
+ttl(#oracle_register_tx{ttl = TTL}) ->
+    TTL.
+
+-spec fee(register_tx()) -> integer().
+fee(#oracle_register_tx{fee = Fee}) ->
+    Fee.
+
+-spec new(map()) -> {ok, register_tx()}.
+new(#{account       := AccountPubKey,
+      nonce         := Nonce,
+      query_spec    := QuerySpec,
+      response_spec := ResponseSpec,
+      query_fee     := QueryFee,
+      ttl           := TTL,
+      fee           := Fee}) ->
+    {ok, #oracle_register_tx{account       = AccountPubKey,
+                             nonce         = Nonce,
+                             query_spec    = QuerySpec,
+                             response_spec = ResponseSpec,
+                             query_fee     = QueryFee,
+                             ttl           = TTL,
+                             fee           = Fee}}.
+
+-spec nonce(register_tx()) -> non_neg_integer().
+nonce(#oracle_register_tx{nonce = Nonce}) ->
+    Nonce.
+
+-spec origin(register_tx()) -> pubkey().
+origin(#oracle_register_tx{account = AccountPubKey}) ->
+    AccountPubKey.
+
+%% Account should exist, and have enough funds for the fee + the query_fee.
+-spec check(register_tx(), trees(), height()) -> {ok, trees()} | {error, term()}.
+check(#oracle_register_tx{account = AccountPubKey, nonce = Nonce,
+                          ttl = TTL, fee = Fee}, Trees, Height) ->
+    Checks =
+        [fun() -> aetx_utils:check_account(AccountPubKey, Trees, Height, Nonce, Fee) end,
+         fun() -> ensure_not_oracle(AccountPubKey, Trees) end,
+         fun() -> aeo_utils:check_ttl_fee(Height, TTL, Fee - ?ORACLE_REGISTER_TX_FEE) end],
+
+    case aeu_validation:run(Checks) of
+        ok              -> {ok, Trees};
+        {error, Reason} -> {error, Reason}
+    end.
+
+-spec signers(register_tx()) -> [pubkey()].
+signers(#oracle_register_tx{account = AccountPubKey}) ->
+    [AccountPubKey].
+
+-spec process(register_tx(), trees(), height()) -> {ok, trees()}.
+process(#oracle_register_tx{account       = AccountPubKey,
+                            nonce         = Nonce,
+                            fee           = Fee} = RegisterTx, Trees0, Height) ->
+    AccountsTree0 = aec_trees:accounts(Trees0),
+    OraclesTree0  = aec_trees:oracles(Trees0),
+
+    Account0 = aec_accounts_trees:get(AccountPubKey, AccountsTree0),
+    {ok, Account1} = aec_accounts:spend(Account0, Fee, Nonce, Height),
+    AccountsTree1 = aec_accounts_trees:enter(Account1, AccountsTree0),
+
+    Oracle = aeo_oracles:new(RegisterTx, Height),
+    OraclesTree1 = aeo_state_tree:enter_oracle(Oracle, OraclesTree0),
```

Comment:
I understand assumption is that oracle is not there. I suggest prefering `aeo_state_tree:insert_oracle` to `aeo_state_tree:insert_interaction` following gb_trees API semantics wrt crash (nice to have, for clarity).

Reply:
If we assume that `check/3` is run before `process/3`, then we have indeed verified that the oracle is not there...

Reply:
Will fix

Reply:
Added insert_oracle and removed enter_oracle since there should be no use for that function.